docker_plugin/utils.py

```diff
@@ -0,0 +1,289 @@
+# coding=utf-8
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+# Third-party Imports
+import docker
+
+# Cloudify Imports
+from cloudify import ctx
+from cloudify.exceptions import RecoverableError, NonRecoverableError
+
+
+def get_image_id(tag, image_id, client):
+    """ Attempts to get the correct image id from Docker.
+
+    :param tag: The image tag provided in the blueprint.
+    :param image_id: The last id extracted from the stream
+        during the image pull process.
+    :param client: docker client
+    returns image id
+        if the last extracted image id matches the tag
+        and the image id for that tag matches the pulled
+        image id then that image id is used
+        if the image id only matches an image id in docker
+        then that is used
+        if no match is found the image id initially passed is
+        kept
+    """
+
+    ctx.logger.debug('This image id {} is an image id stub and '
+                     'was the last image id captured during '
+                     'the pull operation. If you provided a '
+                     'tag, it might not match the image id of '
+                     'the tagged image. So this operation '
+                     'searches for an id that matches that tag. '
+                     'If there is no tag match and only a match '
+                     'between image ids then the complete id is '
+                     'returned. If there is no match at all, the '
+                     'image_id stub is returned.'.format(image_id))
+
+    try:
+        images = client.images()
+    except docker.errors.APIError as e:
+        raise NonRecoverableError('Unable to get last created image: '
+                                  '{}'.format(e))
+
+    for img in images:
+        img_id, tags = (str(img.get('Id')), img.get('RepoTags'))
+        if filter(None, [repotag if
+                  tag in repotag else None for repotag in tags]) \
+                and str(image_id) in img_id:
+            ctx.logger.debug(
+                'The tags and ids match, '
+                'assigning this image id: {}.'.format(img_id))
+            return img_id
+        elif str(image_id) in img_id:
+            ctx.logger.debug(
+                'No tags match, assigning this image id: {}.'.format(img_id))
+            return img_id
+
```

Comment:
I think we should throw a NonRecoverableError here. if the image is not verified then we will not be able to use it anyways and it will crash later on.